chain/bitcoind.go

```diff
@@ -0,0 +1,1050 @@
+				lastHeader = headers.Back().Value.(*btcjson.
+					GetBlockHeaderVerboseResult)
+				lastHash, err = chainhash.NewHashFromStr(
+					lastHeader.Hash)
+				if err != nil {
+					return err
+				}
+			}
+
+			// Check again and make sure we're at the start of the
+			// reorg.
+			if block.Header.PrevBlock.String() != lastHeader.Hash {
+				return errors.New("reorg during rescan went " +
+					"too far back")
+			}
+		}
+
+		// We are at the latest known block, so we notify.
+		lastHeader = &btcjson.GetBlockHeaderVerboseResult{
+			Hash:         block.BlockHash().String(),
+			Height:       i,
+			PreviousHash: block.Header.PrevBlock.String(),
+			Time:         block.Header.Timestamp.Unix(),
+		}
+		blockHash := block.BlockHash()
+		lastHash = &blockHash
+		headers.PushBack(lastHeader)
+
+		_, err = c.filterBlock(block, i, true)
+		if err != nil {
+			return err
+		}
+
+		if i%10000 == 0 {
+			c.onRescanProgress(lastHash, i, block.Header.Timestamp)
+		}
+	}
+
+	return nil
+}
+
+// filterBlock filters a block for watched outpoints and addresses, and returns
+// any matching transactions, sending notifications along the way.
+func (c *BitcoindClient) filterBlock(block *wire.MsgBlock, height int32,
+	notify bool) ([]*wtxmgr.TxRecord, error) {
+	log.Debugf("Filtering block %d (%s)\n", height, block.BlockHash())
+	// Create block details for notifications.
+	blockHash := block.BlockHash()
+	blockDetails := &btcjson.BlockDetails{
+		Hash:   blockHash.String(),
+		Height: height,
+		Time:   block.Header.Timestamp.Unix(),
+	}
+
+	// Cycle through all transactions in the block.
+	var relevantTxs []*wtxmgr.TxRecord
+	for i, tx := range block.Transactions {
+		// Update block and tx details for notifications.
+		blockDetails.Index = i
+		found, rec, err := c.filterTx(tx, blockDetails, notify)
+		if err != nil {
+			log.Warnf("Unable to filter tx: %v", err)
+			continue
+		}
+		if found {
+			relevantTxs = append(relevantTxs, rec)
+		}
+	}
+
+	c.onFilteredBlockConnected(height, &block.Header, relevantTxs)
+
+	c.onBlockConnected(&blockHash, height,
+		block.Header.Timestamp)
+
+	return relevantTxs, nil
+}
+
+// filterTx filters a single transaction against the client's watch list.
+func (c *BitcoindClient) filterTx(tx *wire.MsgTx,
+	blockDetails *btcjson.BlockDetails, notify bool) (bool,
+	*wtxmgr.TxRecord, error) {
+
+	txDetails := btcutil.NewTx(tx)
+	if blockDetails != nil {
+		txDetails.SetIndex(blockDetails.Index)
+	}
+
+	rec, err := wtxmgr.NewTxRecordFromMsgTx(txDetails.MsgTx(), time.Now())
+	if err != nil {
+		log.Errorf("Cannot create transaction record for relevant "+
+			"tx: %v", err)
+		return false, nil, err
+	}
+	if blockDetails != nil {
+		rec.Received = time.Unix(blockDetails.Time, 0)
+	}
+
+	var notifyTx bool
+
+	// If we already know this is a relevant tx from a previous ntfn, we
+	// can shortcut the filter process and let the caller know the filter
+	// matches.
+	if _, ok := c.memPool[tx.TxHash()]; ok {
```

Comment:
I don't see any items from`memPool` being deleted currently. With a _very_ large wallet, this may start to balloon in size. It seems like we can clear this out every few blocks?

Reply:
I still don't get the usage of of this map entirely. 

We add entries to the map for each new block. It includes ever tx within the block. So why does it make sense to notify notifications for a transaction (that may have been sent over the zmq socket as a relevant tx?) immediately, rather than checking to see if it's actually relevant?